### ISSUE TITLE
Dashboard ThemeManager is a singleton service

### ISSUE THREAD

**Issue by tlmii** [dotnet/aspire#2561]
For the local dev scenario this isn't really an issue (though it might be unexpected): A theme change in one browser (not just tab) will impact all browsers.

But for a multi-user scenario, this would be very problematic.

The quick fix here is to change it to Scoped. 

Doing that will be a bit of a take-back, though since then multiple tabs in the same browser for the same user won't get notified. Once the user refreshes the page and picks up the cookie, it'll be fine. But we lose the current behavior of all tabs changing at the same time. We should chat with the Blazor folks and see if there's a common solution for this.

**Comment by JamesNK** [comment]
@SteveSandersonMS I have an interesting scenario for you:

Aspire has a theme changer. We want a theme change to immediately change all browser windows on someone's machine. I think we could achieve that by telling **all** browsers to execute a JS method that looks at some state shared between browser windows (a cookie?) and use it to update the theme. However, that's not efficient because if there are other users of the app then we'd be telling them to update the theme even though they haven't done anything.

Is there a way to identify the Blazor circuits that share browser windows/tabs on someone's machine? If there is then when a theme change happens we can include that ID and get Blazor to notify only browsers for that machine to update.

**Comment by SteveSandersonMS** [comment]
> Is there a way to identify the Blazor circuits that share browser windows/tabs on someone's machine?

There's nothing built-in but you could implement something yourself. For example, assign a GUID to each user in a cookie on their very first HTTP request. That will then be common to all tabs within that browser. You could put some code in one of your components that reads this cookie value using JS interop, and tracks a dictionary of GUID->List<callback>. Then you can look up all the callbacks corresponding to this user.

> all browser windows on someone's machine

I don't think you can know about distinct browsers being on the same machine, as from the server's perspective that's indistinguishable from two different machines behind a NAT. The closest you'll probably get is knowing about distinct connections from te same browser.

**Comment by JamesNK** [comment]
> I don't think you can know about distinct browsers being on the same machine, as from the server's perspective that's indistinguishable from two different machines behind a NAT. The closest you'll probably get is knowing about distinct connections from te same browser.

Yeah, I didn't mean quite this far. All Chrome windows/tabs from the same tab is what I'd like to aim for.

What you suggested is along the lines of what I was thinking of. Thanks for the input!

**Comment by leslierichardson95** [comment]
Closing this issue since it's completed!
